### ISSUE TITLE
"Output Gain" indicator of the Autogain plugin overflows

### ISSUE THREAD

**Issue by hexchain** [wwmm/easyeffects#1441]
### EasyEffects Version

6.2.4

### What package are you using?

Arch (easyeffects)

### Distribution

Arch Linux

### Describe the bug

When the value of Output Gain is too large, the indicator bar overflows:

![image](https://user-images.githubusercontent.com/529520/159152361-12c99473-666a-4892-a973-33f05e212d62.png)

### Expected Behavior

The bar should not overflow.

**Comment by wwmm** [comment]
It happens on my computer too. In my opinion this a GTK4 bug. Even if we are giving a value that is beyond the `GtkLevelBar` limits gtk should not be drawing over other widgets boundaries. Even if I wanted to do that it does not provided functions to do so. It is totally up to gtk to decided where its widgets are drawing. That is why I think this bug is on them.

**Comment by hexchain** [comment]
Opened a issue: https://gitlab.gnome.org/GNOME/gtk/-/issues/4772

**Comment by LebedevRI** [comment]
FWIW https://docs.gtk.org/gtk4/method.LevelBar.set_value.html#parameters says:

value | double
-- | --
  | A value in the interval between GtkLevelBar:min-value and GtkLevelBar:max-value

So it seems like if a value outside of that range is passed, we are in UB lands.

**Comment by wwmm** [comment]
>So it seems like if a value outside of that range is passed, we are in UB lands.

It is still no excuse for gtk4 to be drawing over adjacent widgets. It doesn't make any sense. The drawing has to be contained within the widget boundaries independently of the values we give to it. When I wrote our minimal chart code to draw the spectrum I've noticed that if we are not careful the drawing can be dine over other widgets. This is something that did not happen in gtk3 and that is probably related to the new approach gtk4 has taken about widgets drawing. It is a bug they have to fix.

**Comment by Cyborgscode** [comment]
> It happens on my computer too. In my opinion this is a GTK4 bug. Even if we are giving a value that is beyond the `GtkLevelBar` limits gtk should not be drawing over other widgets boundaries. Even if I wanted to do that it does not provide functions to do so. It is totally up to gtk to decide where its widgets are drawing. That is why I think this bug is on them.

Agreed, A gui toolkit must always check if valid input is supplied. That lesson is typically learned in ALPHA or Beta versions.
But of course, as you now know, that this bug is happening and caused by your input, you can also add a sanity check:

```
if ( value > meter.max ) value = meter.max;
meter.setValue(value);
```

And the next thought is naturally: Did you update the element's maxvalue with your newest "highest possible" value???

```
if ( value > highest ) {
 meter.maxValue(value);
 highest = value;
 }
 meter.setValue(value);
```
it's still the toolkits render bug, but you also did something wrong ;)

**Comment by wwmm** [comment]
>Did you update the element's maxvalue with your newest "highest possible" value???

Some parameters don't really have a defined maximum value. 

>but you also did something wrong ;)

Debatable.

```
if ( value > highest ) {
 meter.maxValue(value);
 highest = value;
 }
```

This isn't a good workaround. Consider the autogain for example. At some point in time you may need a very large gain correction. But a few seconds later the signal may totally change but now that you have set a very large maximum value the level bar is essentially useless. So you have to dynamically decrease the gtklevelbar maximum value back to some small decent value.

Sure it can be done but I do not think it is reasonable that the client has to complicate the code so much for something that is so simple to fix on the toolkit. They have total control of the drawing. Just to do not draw over other toolkits...

**Comment by wwmm** [comment]
I have updated our master branch with a temporary workaround. Setting the GtkLevelBar `overflow` property to `hidden` avoids the drawing over the other widgets. But there is a catch. In the docs for `gtk_widget_set_overflow` there is the following
```
This setting is provided for widget implementations and should not be used by application code
```
So this is something that according to gtk's own docs clients should not be doing because it is the widget implementation duty to care about this. So the proper fix to this is still gtk's responsibility.

**Comment by Cyborgscode** [comment]
someone needs to inform them... so i did:

https://gitlab.gnome.org/GNOME/gtk/-/issues/6352

**Comment by Cyborgscode** [comment]
> This isn't a good workaround. Consider the autogain for example. At some point in time you may need a very large gain correction. But a few seconds later the signal may totally change but now that you have set a very large maximum value the level bar is essentially useless. So you have to dynamically decrease the gtklevelbar maximum value back to some small decent value.

So you never considered an logarithmic axis? Think about it.

**Comment by wwmm** [comment]
>someone needs to inform them... so i did:

There is already an issue opened a long time ago https://gitlab.gnome.org/GNOME/gtk/-/issues/4772. And as you can see here one of its developers were even working on it https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/5835.

>So you never considered an logarithmic axis? Think about it.

It kinda feels like you are here to troll...

**Comment by wwmm** [comment]
As we do not use gtk anymore we can close this issue.
